README.md
# bioC.logs

<!-- badges: start -->
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version-last-release/bioC.logs)](https://cran.r-project.org/package=bioC.logs)
[![cran checks](https://badges.cranchecks.info/worst/bioC.logs.svg)](https://cran.r-project.org/web/checks/check_results_bioC.logs.html)
[![cran checks](https://badges.cranchecks.info/summary/bioC.logs.svg)](https://cran.r-project.org/web/checks/check_results_bioC.logs.html)
[![Downloads](https://cranlogs.r-pkg.org/badges/bioC.logs)](https://cran.r-project.org/package=bioC.logs)
<!--
[![CRAN checks](https://cranchecks.info/badges/worst/bioC.logs)](https://cranchecks.info/pkgs/bioC.logs)
-->
<!-- badges: end -->


## Introduction
This package allows you to download the statistics of BioConductor packages'
downloads as reported by http://bioconductor.org/packages/stats/.


## Usage
The main function of this package is called ```bioC_downloads```.
The function accepts several arguments: `packages names`, `format`, `from/to` **or** `when` and `verbose`.

argument	 | Description
-----------|---------------
`packages names` | is the name(s) of the package(s) for which you want to download the stats from; multiple packages should be indicated using a list of the packages names
`format`     | accepts two options: `"bioC"` (default) will report the downloads as reported by bioconductor, ie. *"Year Month Nb_of_distinct_IPs Nb_of_downloads"*; or, `"CRAN"` will report as CRAN logs does, ie. *"Date  count  package_Name"*
`from/to`    | optional arguments to indicate range of dates to recover the data within -- can NOT be used in combination with `when`
`when`       | optional argument to specify the range of dates to recover the data within -- can NOT be used in combination with `from/to`; possible options are "`ytd`","`year-to-date`","`year-from-now`","`last-year`"
`verbose`    | is a boolean flag indicating whether to print information about the processes
---------------------------


## Features

* The function will return a list containing a dataframe per package entered with columns as indicated by the `format` argument.
Notice that when the `format` is set to `"CRAN"`, the date will be formatted to days-month-year. Because BioConductor reports only totals per month, the "day" in this case will be set to the last date of the corresponding month/year.

* The function will also attempt to report when a package names has been misspelled or just the server is not reachable.
If you are receiving warning messages please check either of these situations.

### Obervations
* The BioConductor website reports total downloads per month.
* The original data also includes an 'all' entry, representing the total downloads per year.
* All the month are reported independently of whether this is the current year and data is still not available.
* You may notice this when requesting the data using the default format from BioConductor, ie. format="bioC".

* When using `format="CRAN"`:
	- the `all` entry is removed
	- the last month to be reported is the previous to the current one, ie. months in the future with no data or the current one (with incomplete data) will not be reported
	- the entries will be order cronologically from the oldest to the newest records


## Installation

For using the "bioC.logs" package, first you will need to install it.

The stable version can be downloaded from the CRAN repository:
```
install.packages("bioC.logs")
```

To obtain the development version you can get it from the github repository, i.e.
```
# need devtools for installing from the github repo
install.packages("devtools")

# install bioC.logs
devtools::install_github("mponce0/bioC.logs")
```

For using the package, either the stable or developmemnt version, just load it
using the library function:
```
# load bioC.logs
library(bioC.logs)
```


## Examples
You will need an active internet connection, as `bioC.logs` will download the
reports from the BioConductor website on demand.

```
# it is possible to download multiple packages, the data will be returned in a list with one entry per package
bioC_downloads(c("ABarray","a4Classif"))

# the 'verbose' option allow you to turn off information reported by the function
bioC_downloads("edgeR",verbose=FALSE)

# setting format="CRAN", will structure the data "a-la-CRAN"
edgeR.logs <- bioC_downloads("edgeR",format="CRAN")
# data is still returned in a list
str(edgeR.logs)
# access data from package
edgeR.logs[[1]]

# get the data in a particular range of dates using 'when'
edgeR.logs <- bioC_downloads("edgeR", when='last-year', format='bioC')

# get the data in a particular range of dates using 'to/from'
# not specifying 'from' will assume since the very first record
edgeR.logs <- bioC_downloads("edgeR", to='03-2015', format='bioC')

# not specifying 'from' will assume until the current day
edgeR.logs <- bioC_downloads("edgeR", from='03-2015', format='bioC')

edgeR.logs <- bioC_downloads("edgeR", from='03-2015',to='05-2016', format='bioC')
```


### Stats
<!-- badges: start -->
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version-last-release/bioC.logs)](https://cran.r-project.org/package=bioC.logs)
[![cran checks](https://badges.cranchecks.info/worst/bioC.logs.svg)](https://cran.r-project.org/web/checks/check_results_bioC.logs.html)
[![cran checks](https://badges.cranchecks.info/summary/bioC.logs.svg)](https://cran.r-project.org/web/checks/check_results_bioC.logs.html)
[![Downloads last.mnth](https://cranlogs.r-pkg.org/badges/bioC.logs)](https://cran.r-project.org/package=bioC.logs)
[![Downloads last.day](https://cranlogs.r-pkg.org/badges/last-week/bioC.logs)](https://cran.r-project.org/package=bioC.logs)
[![Downloads last.day](https://cranlogs.r-pkg.org/badges/last-day/bioC.logs)](https://cran.r-project.org/package=bioC.logs)
<!-- badges: end -->
<p align="center">
	<img src="https://github.com/mponce0/R.pckgs.stats/blob/master/DWNLDS_bioC.logs.png" width="65%" alt="Live download stats" />
	<figcaption>"Live" download stats, figure generated using "Visualize.CRAN.Downloads"</figcaption>
</p>

<object data="https://github.com/mponce0/R.pckgs.stats/blob/master/DWNLDS_bioC.logs.pdf" type="application/pdf" width="700px" height="700px">


NEWS
January 2023: adjusted references to update CRAN badges _ 1.2.1

May 2021: added rmarkdown to list of "suggests" due to knitr changes (Apr. 2021) _ ver 1.2

February 2020: added options for indicating range of dates (to/from & when); adaptive last day of the month; more checks from CRAN format (remove future dates) _ ver 1.1 

February 2020: First release of "bioC.logs" package _ ver 1.0


R/auxs-utils.R
# auxs-utils.R
#  -- M.Ponce


#################################################################################################
##	Auxiliary Utilities file for the bioC.logs package
#
#	Generic auxiliary functions
#
#################################################################################################


### date's aux fns
year.to.date <- function(){
#' function that returns the range dates for the period year-to-date 
#' @keywords internal
	cur.date <- Sys.Date()

	cur.year <- substr(cur.date,1,4)
	return(list(paste(cur.year,"01","01",sep='-'),as.character(cur.date)))
}

lst.year <- function() {
#' function that returns the range dates for the last year 
#' @keywords internal
        cur.date <- Sys.Date()

        last.year <- as.integer(substr(cur.date,1,4))-1
        return(list(paste(last.year,"01","01",sep='-'),paste(last.year+1,"01","01",sep='-')))

}

year.from.now <- function() {
#' function that returns the date from one year ago
#' @keywords internal
        cur.date <- Sys.Date()
        cur.year <- substr(cur.date,1,4)
        lst.year <- as.integer(cur.year) - 1
        t0 <- paste(lst.year,substr(cur.date,5,10),sep="")

        return(list(t0,cur.date))
}

today <- function() {
#' function that returns the current date 
#' @keywords internal
	t1 <- Sys.Date()
	# Will substract 1 to do not consider today, but yesterday
	# this matches the definition from cran_downloads too
	t1 <- t1 - 1 
	message("Ending date was not specifed, will assume today: ",t1)
	return(format(t1,format="%d-%m-%Y"))
}

origin.of.times <- function() {
#' function that provides a "beginning of times" default date
#' @keywords internal
	t0 <- as.Date("1980-01-01",format="%Y-%m-%d")
	message("Initial date was not specified, will assume:", t0)
	return(format(t0,format="%d-%m-%Y"))
}



Xyear.from.now <- function() {
#' function that returns the date from one year ago
#' @keywords internal
	cur.date <- Sys.Date()-1
	cur.year <- substr(cur.date,1,4)
	lst.year <- as.integer(cur.year) - 1
	t0 <- paste(lst.year,substr(cur.date,5,10),sep="")

	message("Starting date was not specified, will assume a year from now: ",t0)
	return(t0)
}

checkDates <- function(t0,t1) {
#' function to check dates, ie that t0<t1 and t0!=t1
#' @param  t0  initial date
#' @param  t1  final date
#' @return a list with t0 being [[1]] and t1 being [[2]]
#'
#' @keywords internal

	# check if there is any missing date
	if (is.null(t0)) t0 <- origin.of.times()
	if (is.null(t1)) t1 <- as.Date(today(),format="%d-%b-%Y")

	# check whether t0 is greater than t1
	if (as.Date(t0) > as.Date(t1)) {
		# flip dates, t0 will be set to the older date
		ttemp <- t0
		t0 <- t1
		t1 <- ttemp
	} else if (as.Date(t0) == as.Date(t1)) {
		# dates should be different
		stop(t0," and ", t1," should be different!")       
	}

	return(list(t0,t1))
}

checkDate <- function(date.candidate) {
#' function to check date format for "MM-YYYY"
#' @param  date.candidate  date candidate
#' 
#'
#' @keywords internal

	# bisecting month and year...
	MM <- as.integer(substr(date.candidate,1,2))
	YYYY <- as.integer(substr(date.candidate,4,7))

	if ( !is.numeric(MM) |  ((MM<1) | (MM>12)) )
		stop(paste("Problem detected with date: ",date.candidate,"\n","Dates should be specified in 'MM-YYYY' format, with MM from '01' to '12'"))

	if (!is.numeric(YYYY))
		stop(paste("Problem detected with date: ",date.candidate,"\n","Dates should be specified in 'MM-YYYY' format, with MM from '01' to '12'"))
}
	

checkValidDates <- function(t0,t1) {
#' function to check dates, ie that t0<t1 and t0 and t1 are in MM-YYYY format
#' @param  t0  initial date
#' @param  t1  final date
#' @return a list with t0 being [[1]] and t1 being [[2]]
#'
#' @keywords internal

	# check date structure for t0 & t1
	if (!is.null(t0)) {
		checkDate(t0)
	} else {
		t0 <- substr(origin.of.times(),4,10)
	}
	if (!is.null(t1)) {
		checkDate(t1)
	} else {
		t1 <- substr(today(),4,10)
	}

	fake.t0 <- as.Date(paste0("01-",t0),format='%d-%m-%Y')
	fake.t1 <- as.Date(paste0("01-",t1),format='%d-%m-%Y')
	#fake.t1 <- as.Date(paste0(last.day.month(

	# check candidate dates
	finalDates <- checkDates(fake.t0,fake.t1)

        #return(list(t0,t1))
	return(finalDates)
}



last.day.month <- function(MonthYear) {
#' function that returns the number of dates in a given month-year
#' @param  MonthYear  string containing month and year, ie. MM-YYYY
#' @return number of days --as characters-- in the given month-year
#'
#' @keywords internal

	# convert year to numeric
	year <- as.numeric(substr(MonthYear,5,8))
	month <- substr(MonthYear,1,3)

	# check for leap years
	leap = FALSE
	if ( (year %% 4 == 0 ) | ( year %% 100 == 0  &  year %% 400 == 0) )
		leap <-  TRUE

	# obtain number of days
	nbr.days <- switch(tolower(month),
			'jan' = 31,
			'feb' = 28 + leap,  # adds 1 if leap year
                	'mar' = 31,
                	'apr' = 30,
                	'may' = 31,
                	'jun' = 30,
                	'jul' = 31,
                	'aug' = 31,
                	'sep' = 30,
                	'oct' = 31,
                	'nov' = 30,
                	'dec' = 31 )

	return(as.character(nbr.days))
}


daysInMonth <- function(d = Sys.Date()){
#' function that returns the number of dates in a given date
#' @param  d  string containing a date in YYYY-MM-DD format
#' @return number of days in the given the date 
#'
#' @keywords internal

  m = substr((as.character(d)), 6, 7)              # month number as string
  y = as.numeric(substr((as.character(d)), 1, 4))  # year number as numeric

  # Quick check for leap year
  leap = 0
  if ((y %% 4 == 0 & y %% 100 != 0) | y %% 400 == 0)
    leap = 1

  # Return the number of days in the month
  return(switch(m,
                '01' = 31,
                '02' = 28 + leap,  # adds 1 if leap year
                '03' = 31,
                '04' = 30,
                '05' = 31,
                '06' = 30,
                '07' = 31,
                '08' = 31,
                '09' = 30,
                '10' = 31,
                '11' = 30,
                '12' = 31))
}



##### //////////////////////////////////////////////////////////////////////////// #####
##### //////////////////////////////////////////////////////////////////////////// #####
##### //////////////////////////////////////////////////////////////////////////// #####



R/bioC_logs.R
# Main Downloading 'general' Engine for "bioC.logs"
#
# -- M.Ponce

####################################################################################
####################################################################################
#########################     MAIN DOWNLOADING ENGINE     ##########################
####################################################################################
####################################################################################

############################################################################
############################################################################
#####  L E V E L   _0_   #####   MASTER  DOWNLOADING  ENGINE   #############
############################################################################
############################################################################

bioC_download_engine <- function(pckg=NULL, bioC.categ='bioc', target='stats', format="bioC", from=NULL,to=NULL, when=NULL, verbose=TRUE) {
#' main downloading engine to download logs from BioConductor stats
#' @param  pckg  list of packages names
#' @param  bioC.categ  one of the three categories of packages in BioConductor: "bioc", "workflows" or "data-experiment"
#' @param  target  one of the two reported "scores" in BioConductor: "stats" (ie. nbr of downloads) or "scores"
#' @param  format  two options: "bioC" (default) will report the downloads as reported by BioConductor, ie. "Year  Month  Nb_of_distinct_IPs  Nb_of_downloads"; or, "CRAN" will report as CRAN logs does, ie. "Date  count  package_Name" (in cranlogs 'Nb_of_downloads' are referred as 'count')
#' @param  from  date in "MM-YYYY" format, specifying the initial date to be considered (optional argument)
#' @param  to  date in "MM-YYYY" format, specifying the final date to be considered (optional argument)
#' @param  when  optional argument, to specify pre-defined range dates; ie. 'ytd', 'year-to-date', 'last-year'
#' @param  verbose  boolean flag indicating whether to print information about the processes...
#'
#' @return a list containing a dataframe per package entered with columns as indicated by the format argument
#'
#' @importFrom utils  read.table
#' @importFrom stats  na.omit
#' @export
#'
#' @examples
#' bioC_download_engine(c("ABarray","a4Classif"))
#' bioC_download_engine("edgeR", target="scores", verbose=FALSE)
#' edgeR.logs <- bioC_download_engine("edgeR", bioC.categ="bioc", format="CRAN")
#' fluentGenomics.logs <- bioC_download_engine("fluentGenomics", bioC.categ='workflows',
#'						when='last-year', format='bioC')
#' edgeR.logs <- bioC_download_engine("edgeR", to='03-2015', format='bioC')
#' pasilla.logs <- bioC_download_engine("pasilla", bioC.categ='data-experiment', target='scores',
#'					from='03-2015', format='bioC')
#' edgeR.logs <- bioC_download_engine("edgeR", from='03-2015',to='05-2016', format='bioC')

	## function to inform the user about some general observations
	bioC_disclaimer <- function() {
		header <- paste(paste(rep("-",80),collapse=''), '')
		msg <- paste(header,
			"The BioConductor website reports total downloads per month.",
			"The data also includes an 'all' entry, representing the total downloads per year.",
			"All the month are reported independently of whether this is the current year and data is still not available.",
			"You may notice this when requesting the data using the default format from BioConductor.",
			header, sep='\n')

		message(msg)
	}


	## function for error handling
	errorHandling.Msg <- function(condition,pck) {
		message("A problem was detected when trying to retrieve the data for the package: ",pck)
		if (grepl("404 Not Found",condition)) {
			message("It is possible that you misspeled the name of this package! Please check!")
		} else {
			message("It is possible that your internet connection is down! Please check!")
		}
		message(condition,'\n')

		# update problems counter
		pkg.env$problems <- pkg.env$problems + 1
	}

	###############################

	# define valid formats
	fmt.opts <- c("bioC","CRAN")

	# define possible BioConductor categories: "bioc", "workflows", or "data-experiment"
	valid.BioCcategs <- c("bioc", "workflows", "data-experiment")

	# define possible BioConductor stats: 'stats' or 'scores'
	valid.BioCtargets <- c("stats", "scores")
	# define valid TOTAL KEYS
	valid.BioCtotals <- c("bioc_pkg","workflows_pkg","experiment_pkg")

	# define date ranges for 'when'
	when.opts <- c("ytd","year-to-date","last-year","year-from-now")
	date.range <- NULL

	# CHECKS
	# check that bioC.categ is a valid option
	if (!(bioC.categ %in% valid.BioCcategs)) stop("Invalid ", bioC.categ,"! Valid options are:", paste(valid.BioCcategs,collpase=' '))
	# check that target is a valid option
	if (!(target %in% valid.BioCtargets)) stop("Invalid ", target,"! Valid options are:", paste(valid.BioCtargets,collapse=' '))
	# checks for TOTALS...
	going.for.totals <- FALSE
	if ( length(pckg) == 1 )  {
		if (pckg %in% valid.BioCtotals) going.for.totals <- TRUE
	}

	# Define BioConductor URL and file ending
	conjuction.url <- "stats"
	bioC.url <- paste0("http://bioconductor.org/packages/",conjuction.url,"/",bioC.categ,"/")
	ending <- paste0("_",target,".tab")

	# define day to when assign the value of downloads (BioConductor stats report monthly values)
	# will pcik '28' as last day of the month, to avoid issues with Feb...
	# day.assign <- "28"

	# initialize container for results
	pckgs.stats <-c() 


	#######

	# check valid argument
	if (is.null(pckg) || !is.character(pckg)) {
		stop("Must specify a valid package name!")
	}

	# check valid format
	if (!is.character(format) | !(format %in% fmt.opts) ) {
		stop("Invalid 'format'! Possible optiosn are: ", paste(fmt.opts,collapse=" "))
	}

	# check valid dates
	if ( (!is.null(from) | !is.null(to))  &  !is.null(when) )
		stop("Options 'from'/'to' and 'when' can not be combined. \n You either must specify a range 'from'-'to' or use 'when'.")

	# check whether the argument 'when' has been specified
        if (!is.null(when)) {
		if (tolower(when) %in% when.opts) {
			if (tolower(when) == "ytd" | tolower(when) == "year-to-date") {
				date.range <- year.to.date()
			} else if (tolower(when) == "last-year") {
				date.range <- lst.year()
			} else if (tolower(when) == "year-from-now") {
				date.range <- year.from.now()
			}
		} else {
			message("Unrecognized option for argument 'when', valid ones are: ", paste(when.opts,collapse=" "))
		}
	}

	# check whether 'from/to' was specified
	if (!is.null(from) | !is.null(to)) {
		date.range <- checkValidDates(from,to)
	}

	# if format was set to 'CRAN' and not date range was specified, set it to origin and current date
	if (format=="CRAN" & is.null(from) & is.null(to) & is.null(when)) {
		date.range <- checkValidDates(NULL,NULL)
	}


	# check that verbose is a boolean
	if (!is.logical(verbose)) {
		stop("'verbose' should be a Boolean variable, i.e. TRUE or FALSE!")
	}
	#######


	# Counter for detection or problems, defined within the pckg environ to avoid global variables, ie. <<- 
	pkg.env <- new.env()
	pkg.env$problems <- 0

	# process package list
	for (ind.pckg in seq_along(pckg)) {
		pck <- pckg[ind.pckg]

		pckgFile <- paste0(pck,'/',pck,ending)
		if (going.for.totals) pckgFile <- paste0(pck,ending)
		pckg.URL <- paste0(bioC.url,pckgFile)

		# Attempt to protect against bad internet conenction or misspelled package name
		tryCatch(
			{
			# read data
			pckg.data <- read.table(pckg.URL, header=TRUE)

			# this implies that a range of dates have been specified
			#if (!is.null(when) | (!is.null(from) & !is.null(to)) ) {
if(!is.null(date.range)) {
			# check whether the argument 'when' has been specified
#			if (!is.null(when)) {
#				if (tolower(when) %in% when.opts) {
#					if (tolower(when) == "ytd" | tolower(when) == "year-to-date") {
#						date.range <- year.to.date()
#					} else if (tolower(when) == "last-year") {
#						date.range <- lst.year()
#					} else if (tolower(when) == "year-from-now") {
#						date.range <- year.from.now()
#					}
				# hence will have to remove the 'all'
				pckg.data <- pckg.data[!as.character(pckg.data$Month)=="all",]

					# restrict subset to specified dates...
					# will assign the last day of the month for each specific entry
					day.assign <- sapply(paste(pckg.data$Month,pckg.data$Year,sep='-'),last.day.month)
					pckg.dates <- paste(pckg.data$Year,pckg.data$Month,day.assign,sep='-')
					#.#.#pckg.dates <- paste(pckg.data$Year,pckg.data$Month,last.day.month(pckg.data$Month,pckg.data$Year), sep='-')
					# re-format dates to match YYYY-mm-dd
					pckg.dates <- format(as.Date(pckg.dates, format = "%Y-%b-%d"), "%Y-%m-%d")
					# select the entries within the actual range of dates
					# na.omit() is necessary as bioC-logs include totals per month labelled "all"
					pckg.data <- na.omit(pckg.data[((pckg.dates > date.range[[1]]) & (pckg.dates < date.range[[2]])),])
#				} else {
#					message("Unrecognized option for argument 'when', valid ones are: ", paste(when.opts,collapse=" "))
#				}
			}


			# reformat data into CRAN format
                        if (format=="CRAN") {
                                if (ind.pckg==1 && verbose) message("Data will be returned as 'date  count  packageName'")

                                # clean the entrie 'all' totals per year...
                                clean.data <- pckg.data[!as.character(pckg.data$Month)=="all",]
				# set the date to the last day of the month
				day.assign <- sapply(paste(clean.data$Month,clean.data$Year,sep='-'),last.day.month)
                                pckg.dates <- paste(clean.data$Year,clean.data$Month,day.assign,sep='-')

				#new.df <- data.frame( date=as.Date(paste0(day.assign,clean.data$Month,clean.data$Year), "%d%b%Y"),
				new.df <- data.frame( date=as.Date(pckg.dates, '%Y-%b-%d'),
				#.#.#new.df <- data.frame( date=as.Date(paste0(last.day.month(clean.data$Month,clean.data$Year),clean.data$Month,clean.data$Year), "%d%b%Y"),
							# in cranlogs 'Nb_of_downloads' are referred as 'count'
							count=as.numeric(clean.data$Nb_of_downloads),
							package=pck )

				# in cranlogs, also the dates are ordered...
				new.df <- new.df[order(as.Date(new.df$date,format='%Y-%b-%d')),]
                                pckg.data <- new.df
                        }

			pckgs.stats[[ind.pckg]] <- pckg.data 
			},

			# warning
			warning = function(cond) {
					errorHandling.Msg(cond,pck)
				},
			# error
			error = function(e){
					errorHandling.Msg(e,pck)
				}
			)
	}

	# final report
	if (verbose) {
		message(paste(ind.pckg-pkg.env$problems)," packages processed from a total of ",ind.pckg," requests!")
		if ((ind.pckg-pkg.env$problems) != 0) {
			message("Data was retrieved from ",bioC.url," using the *",format,"* format.")
			if (verbose) bioC_disclaimer()
		}

		# problems report
		if (pkg.env$problems != 0)
			message(pkg.env$problems," problems detected, the associated entry will be set to NULL")
	}

	# return results
	return(pckgs.stats)
}

####################################################################################
####################################################################################

##### //////////////////////////////////////////////////////////////////////////// #####

